[Merged by Bors] - feat(algebra/lie/basic): define ideal operations for Lie modules

ocfnash opened the pull request:
---
I am hoping that if these changes are merged, I can use them for https://github.com/leanprover-community/mathlib/pull/5241

Commit:
feat(algebra/lie/basic): define ideal operations for Lie modules

Commit:
Fix ambiguous reference `mk_coe`

jcommelin commented on src/algebra/lie/basic.lean:
```diff
@@ -774,7 +781,9 @@ instance : partial_order (lie_submodule R L M) :=
-lemma le_def (N N' : lie_submodule R L M) : N ≤ N' ↔ (N : set M) ⊆ N' := iff.rfl
+lemma le_def : N ≤ N' ↔ (N : set M) ⊆ N' := iff.rfl
+
+lemma le_def_submodule : N ≤ N' ↔ (N : submodule R M) ≤ N' := iff.rfl
```
If you swap the two sides (and call it something like `coe_le_coe_submodule`, then it would make a useful simp-lemma.

jcommelin commented on src/algebra/lie/basic.lean:
```diff
@@ -891,6 +926,84 @@ end lie_span
+instance : has_bracket (lie_ideal R L) (lie_submodule R L M) :=
+⟨λ I N, lie_submodule.lie_span R L { m | ∃ (x : I) (n : N), ⁅(x : L), (n : M)⁆ = m }⟩
+
+lemma lie_apply :
```
why exactly this name?

ocfnash replied:
A moment of madness? This is obviously a terrible name. I have just pushed a commit renaming it to `lie_ideal_oper_eq_span` which seems better but still not ideal.

jcommelin commented on src/algebra/lie/basic.lean:
```diff
@@ -891,6 +926,84 @@ end lie_span
+lemma lie_mem_lie (x : I) (m : N) : ⁅(x : L), (m : M)⁆ ∈ ⁅I, N⁆ :=
+by { rw lie_apply, apply subset_lie_span, use [x, m], }
+
+lemma lie_le : ⁅I, N⁆ ≤ N :=
```
I think you'll also want `[I, J] \le I` when you specialise to ideals, so these should probably be called `lie_le_right` and `lie_le_left`.
```suggestion
lemma lie_le_right : ⁅I, N⁆ ≤ N :=
```

ocfnash replied:
Great point!

jcommelin reviewed (commented):
Great. These will be helpful later on.

Commit:
Suggestions from code review

Commit:
Rename 'lie_apply` --> `lie_ideal_oper_eq_span`

Alternative suggestions welcome: struggling to settle on a good name!

Commit:
Simplify proof of `lie_le_inf`

Now that we have both `lie_le_right` and `lie_le_left` we can shorten this proof.

eric-wieser commented on src/algebra/lie/basic.lean:
```diff
@@ -718,6 +718,13 @@ instance has_mem : has_mem M (lie_submodule R L M) := ⟨λ x N, x ∈ (N : set
 iff.rfl
 
+@[simp] lemma mk_coe (S : set M) (h₁ h₂ h₃ h₄) :
+  ((⟨S, h₁, h₂, h₃, h₄⟩ : lie_submodule R L M) : set M) = S := rfl
```
I'm surprised we don't have this lemma for any of the other sub_* structures.

ocfnash replied:
Me too.

I believe it is just one of those things we haven't really needed yet but I'd love to be corrected if there is a good reason not to have these lemmas.

Vierkantor replied:
I think it is exactly that. Also the total number of lemmas you would have to add is quadratic in the number of "extends" (`coe_to_set_mk`, `coe_to_submodule_mk`, `coe_to_subalgebra_mk`, ...), so we should probably automate it at some point.

Vierkantor commented on src/algebra/lie/basic.lean:
```diff
@@ -718,6 +718,13 @@ instance has_mem : has_mem M (lie_submodule R L M) := ⟨λ x N, x ∈ (N : set
 iff.rfl
 
+@[simp] lemma mk_coe (S : set M) (h₁ h₂ h₃ h₄) :
+  ((⟨S, h₁, h₂, h₃, h₄⟩ : lie_submodule R L M) : set M) = S := rfl
```
Since the `coe` arrow comes on the right of the `mk`, and to avoid confusion with the `coe_mk {S : subobject X} (x : X) (hx : x ∈ S) : (subobject.mk x hx : X) = x` lemmas (I'm not sure that `lie_submodule` has that one already), we should call this:
```suggestion
@[simp] lemma coe_to_set_mk (S : set M) (h₁ h₂ h₃ h₄) :
  ((⟨S, h₁, h₂, h₃, h₄⟩ : lie_submodule R L M) : set M) = S := rfl
```

Vierkantor commented on src/algebra/lie/basic.lean:
```diff
@@ -718,6 +718,13 @@ instance has_mem : has_mem M (lie_submodule R L M) := ⟨λ x N, x ∈ (N : set
+
+@[simp] lemma mk_coe_submodule (p : submodule R M) (h) :
+  (({lie_mem := h, ..p} : lie_submodule R L M) : submodule R M) = p :=
+by { cases p, refl, }
```
```suggestion
@[simp] lemma coe_to_submodule_mk (p : submodule R M) (h) :
  (({lie_mem := h, ..p} : lie_submodule R L M) : submodule R M) = p :=
by { cases p, refl, }
```

Vierkantor commented on src/algebra/lie/basic.lean:
```diff
@@ -774,7 +781,9 @@ instance : partial_order (lie_submodule R L M) :=
-lemma le_def (N N' : lie_submodule R L M) : N ≤ N' ↔ (N : set M) ⊆ N' := iff.rfl
+lemma le_def : N ≤ N' ↔ (N : set M) ⊆ N' := iff.rfl
+
+@[simp] lemma coe_le_coe_submodule : (N : submodule R M) ≤ N' ↔ N ≤ N' := iff.rfl
```
This should definitely be a `norm_cast` lemma. Making the name more symmetrical, as in the suggestion, would be nice, if you don't feel it's too long.
```suggestion
@[simp, norm_cast] lemma coe_submodule_le_coe_submodule : (N : submodule R M) ≤ N' ↔ N ≤ N' := iff.rfl
```

Vierkantor commented on src/algebra/lie/basic.lean:
```diff
@@ -840,7 +849,34 @@ instance : add_comm_monoid (lie_submodule R L M) :=
-@[simp] lemma add_eq_sup (N N' : lie_submodule R L M) : N + N' = N ⊔ N' := rfl
+@[simp] lemma add_eq_sup : N + N' = N ⊔ N' := rfl
+
+lemma sup_eq_sup_submodule :
```
About my suggestion: making it `norm_cast` is definitely a good idea, changing the name I'm not totally convinced by.
```suggestion
@[norm_cast] lemma coe_sup :
```

Vierkantor commented on src/algebra/lie/basic.lean:
```diff
@@ -986,6 +1104,32 @@ end quotient
+namespace lie_module
+
+/-- The derived Lie submodule of a Lie module. -/
+def derived_lie_submodule : lie_submodule R L M := ⁅(⊤ : lie_ideal R L), ⊤⁆
```
The full name looks like `lie_module.derived_lie_submodule`. Maybe this should just be `lie_submodule.derived`?

ocfnash replied:
I'm happy to make a change but I'm a bit unsure about `lie_submodule.derived`.

This is the derived Lie submodule of a Lie module. Am I silly to think that the bit to the left of the `.` is the thing from which we derive and the bit to the right is the thing we get when we derive?

How would you feel about something in the root namespace called `derived_lie_submodule`?

Vierkantor replied:
My motivation was that you could read `lie_submodule.derived` as "the `lie_submodule` which is `derived`". But `derived_lie_submodule` in the root namespace sounds fine too :+1:

ocfnash replied:
I've gone for `derived_lie_submodule` in the root.

Vierkantor reviewed (changes requested):
Some style suggestions. I'm not very familiar with Lie algebra in mathlib, so I have the feeling `sup_eq_sup_submodule` can be golfed a bit but haven't tried doing so yet.

Commit:
Update src/algebra/lie/basic.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

Commit:
Update src/algebra/lie/basic.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

Commit:
Update src/algebra/lie/basic.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

Commit:
Update src/algebra/lie/basic.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

Commit:
Catch up after code review suggestions

jcommelin committed:
minigolf

jcommelin reviewed (approved):
LGTM

Commit:
Improve derived Lie submodule naming

Commit:
Simplify `lie_submodule.coe_sup`

Using `Inf_le_Inf` (which I originally introduced) actually makes things
more work.

Vierkantor reviewed (approved):
Thanks :tada:

bors r+
